docs/design/readme.md

```diff
@@ -0,0 +1,1704 @@
+    /// but that would be poorly compatible with listen-only nodes that are not interested in sending
+    /// any transfers, as they require no Tx sockets at all.
+    virtual [[nodiscard]]
+    std::expected<UniquePtr<ITxSocket>,
+                  std::variant<MemoryError, PlatformError>>
+    makeTxSocket() = 0;
+
+    /// Construct a new socket bound to the specified multicast group endpoint.
+    virtual [[nodiscard]]
+    std::expected<UniquePtr<IRxSocket>,
+                  std::variant<MemoryError, PlatformError, ArgumentError>>
+    makeRxSocket(const UDPIPEndpoint& multicast_endpoint) = 0;
+};
+```
+
+If `ITxSocket::send` returns false, indicating that the socket is not ready to accept writes, the implementation would normally register this occurrence internally and then register the socket's file descriptor for writing via `IPollSet::output` at the next `IRunnable::run` or via `IMultiplexer::add,` depending on which proposal is chosen. This behavior ensures that write polling will only be performed if there are pending transmissions. In the first case (the stateless case with `IPollSet`), one potential limitation to be aware of is that in the following call sequence, the pending transmission becomes delayed by one `run` cycle:
+
+- Initially, no pending transmission is registered (no data to send).
+- `run` is called, but no pending transmission is registered, so `output` is not called.
+- `send` is called and the socket is found to be not ready to accept writes. Output polling is needed.
+- The main loop is blocked on IO until the next `run` service cycle. However, the IO multiplexor is unaware of the fact that there is data waiting to be sent, which results in an unnecessary transmission delay.
+
+The second case (the stateful case) does not have such limitations as it allows the multiplexer to modify the poll set on the fly.
+
+```c++
+/// The IRunnable interface allows the higher layers to retrieve the IO descriptors for polling.
+class ITxSocket : public IRunnable
+{
+public:
+    /// Returns true if the frame has been accepted successfully, false if the socket is not ready for writing.
+    /// The payload may be fragmented to minimize data copying in the user space, allowing the implementation to
+    /// use vectorized I/O (iov).
+    virtual [[nodiscard]] std::expected<bool, std::variant<PlatformError, ArgumentError>>
+    send(const UDPIPEndpoint&                              multicast_endpoint,
+         const std::uint8_t                                dscp,
+         const std::span<const std::span<const std::byte>> payload_fragments) = 0;
+
+    /// This is the Cyphal-layer MTU, which is the maximum number of payload bytes per Cyphal frame.
+    /// To guarantee a single frame transfer, the maximum payload size shall be 4 bytes less to accommodate the CRC.
+    /// The MTU is set per Tx socket individually and it must remain constant after initialization.
```

Comment:
how will you handle changes to the system while running?

Reply:
This design assumes that the MTU of the socket does not change while it's open. We could remove this assumption and query the MTU before every transmission through the socket.

Reply:
This is a more general problem we'll need to plan for. On non-embedded platforms interfaces can come and go, can change their configuration, and generally be chaotic. Trying to handle every possible change specifically is probably not tenable. I imagine we could have some sort of "reconfigure on change" behaviour where, for any given change that affects established parameters (e.g. MTU) we simply repeat the initialization logic used to acquire the media resource in the first place.

Reply:
The redundant interface class is helpful for runtime reconfiguration